### PR TITLE
Ceph: integration test creates ObjectStoreUser before ObjectStore

### PR DESCRIPTION
**Pull request by umangachapagain**
Signed-off-by: Umanga Chapagain <chapagainumanga@gmail.com>

**Description of your changes:**
This fix is added to ensure that everything works as expected
when user tries to create ObjectStoreUser before creating
ObjectStore itself. ObjectStoreUser will wait for ObjectStore
to be up and running.

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

[test ceph]
[test full]

### CODE REVIEW

**Comment by leseb** [comment]
1.13 and 1.17 previously failed.

**Comment by umangachapagain** [comment]
> 1.13 and 1.17 previously failed.

Failure on 1.17 is as expected. Looks like the tests on all other versions were simply skipped.

When creating ObjectUser before ObjectStore for tests, for some reason rgw pod does not come up and I have been trying to figure out why.

**Comment by leseb** [comment]
> > 1.13 and 1.17 previously failed.
> 
> Failure on 1.17 is as expected. Looks like the tests on all other versions were simply skipped.
> 
> When creating ObjectUser before ObjectStore for tests, for some reason rgw pod does not come up and I have been trying to figure out why.

Why is 1.17 failure expected?

**Comment by umangachapagain** [comment]
> > > 1.13 and 1.17 previously failed.
> > 
> > 
> > Failure on 1.17 is as expected. Looks like the tests on all other versions were simply skipped.
> > When creating ObjectUser before ObjectStore for tests, for some reason rgw pod does not come up and I have been trying to figure out why.
> 
> Why is 1.17 failure expected?

Because we hit this issue multiple times on multiple PRs during CI test. So, this PR was to reproduce the issue, pin point it and fix whatever needs to be fixed. (Because we don't know what's going wrong)

And, since we do allow objectUser to be created before objectStore it is required that the CI passes for this PR and is a follow-up for https://github.com/rook/rook/pull/4149#issuecomment-550960587

**Comment by leseb** [comment]
> > > > 1.13 and 1.17 previously failed.
> > > 
> > > 
> > > Failure on 1.17 is as expected. Looks like the tests on all other versions were simply skipped.
> > > When creating ObjectUser before ObjectStore for tests, for some reason rgw pod does not come up and I have been trying to figure out why.
> > 
> > 
> > Why is 1.17 failure expected?
> 
> Because we hit this issue multiple times on multiple PRs during CI test. So, this PR was to reproduce the issue, pin point it and fix whatever needs to be fixed. (Because we don't know what's going wrong)
> 
> And, since we do allow objectUser to be created before objectStore it is required that the CI passes for this PR and is a follow-up for [#4149 (comment)](https://github.com/rook/rook/pull/4149#issuecomment-550960587)

If Ceph is unhealthy (unknown PGs) that could block the user's creation.

**Comment by umangachapagain** [comment]
> 
> If Ceph is unhealthy (unknown PGs) that could block the user's creation.
>
How do we ensure that's not the case during tests?
Pg Autoscaler is enabled. And, we don't do anything different with the cluster.

Also, why would it happen only when I try to test creation of ObjectUser before ObjectStore and not the other way around. (ObjectUser will wait for the ObjectStore to come up which makes it technically the same flow)

**Comment by leseb** [comment]
> > If Ceph is unhealthy (unknown PGs) that could block the user's creation.
> 
> How do we ensure that's not the case during tests?
> Pg Autoscaler is enabled. And, we don't do anything different with the cluster.
> 
> Also, why would it happen only when I try to test creation of ObjectUser before ObjectStore and not the other way around. (ObjectUser will wait for the ObjectStore to come up which makes it technically the same flow)

IIRC it's not possible to create the user if not rgw is up and running since the user is stored in a rgw pool.
Based on that, it makes sense that we see this working when the user is created after the object store.

**Comment by umangachapagain** [comment]
> Have you tested locally that you can create a ceph object user before defining the object store CR? Or do you see this error locally as well?
> 
On 1st try I did hit this error. When I tried 2nd and 3rd time, it worked fine.
Investigating this as it looks like more than just an Integration Test issue.

**Comment by umangachapagain** [comment]
> What's the status on this PR? Thanks

Creating ObjectStoreUser before ObjectStore seems to be working locally.
Still unable to figure out why Integration Test would fail.

**Comment by leseb** [comment]
@umangachapagain can you rebase and see how the CI reacts? Thanks!

**Comment by umangachapagain** [comment]
> @umangachapagain can you rebase and see how the CI reacts? Thanks!

Interesting. The tests have passed.
Can you re-run the CI ones just to ensure this isn't a flake?

**Comment by leseb** [comment]
> > @umangachapagain can you rebase and see how the CI reacts? Thanks!
> 
> Interesting. The tests have passed.
> Can you re-run the CI ones just to ensure this isn't a flake?

I think my recent changes of separating the object controller and object user controller to controller runtime helped :).
Sure, ran again.

**Comment by leseb** [comment]
Well `[test full]` cannot be flake as **everything** runs on every kube version. So I think we should go ahead with that change.

File: tests/integration/ceph_base_object_test.go [hunk 1]
```diff
@@ -50,20 +50,31 @@ func runObjectE2ETest(helper *clients.TestClient, k8sh *utils.K8sHelper, s suite
 	logger.Infof("Object Storage End To End Integration Test - Create Object Store, User,Bucket and read/write to bucket")
 	logger.Infof("Running on Rook Cluster %s", namespace)
 
-	logger.Infof("Step 0 : Create Object Store")
+	logger.Infof("Step 0 : Create Object Store User")
+	cosuErr := helper.ObjectUserClient.Create(namespace, userid, userdisplayname, storeName)
+	require.Nil(s.T(), cosuErr)
+
+	time.Sleep(5 * time.Second)
```
Comment (leseb) [inline tests/integration/ceph_base_object_test.go]: can we remove this now?

Comment (umangachapagain) [reply tests/integration/ceph_base_object_test.go]: Done.

**Pull request merged**